Use `__builtin_setjmp` instead of `sigsetjmp`.

sunfishcode opened the pull request:
Use [`__builtin_setjmp`] instead of `sigsetjmp`, as it is implemented in
the compiler, performed inline, and saves much less state. This speeds up
calls into wasm by about 8% on my machine.

[`__builtin_setjmp`]: https://gcc.gnu.org/onlinedocs/gcc/Nonlocal-Gotos.html

<!--

Please ensure that the following steps are all taken care of before submitting
the PR.

- [ ] This has been discussed in issue #..., or if not, please tell us why
  here.
- [ ] A short description of what this does, why it is needed; if the
  description becomes long, the matter should probably be discussed in an issue
  first.
- [ ] This PR contains test cases, if meaningful.
- [ ] A reviewer from the core maintainer team has been assigned for this PR.
  If you don't know who could review this, please indicate so. The list of
  suggested reviewers on the right can help you.

Please ensure all communication adheres to the [code of
conduct](https://github.com/bytecodealliance/wasmtime/blob/master/CODE_OF_CONDUCT.md).
-->

bjorn3 commented on crates/runtime/src/helpers.c:
```diff
@@ -2,18 +2,22 @@
+// [clang docs]: https://llvm.org/docs/ExceptionHandling.html#llvm-eh-sjlj-setjmp
+#define platform_setjmp(buf) __builtin_setjmp(buf)
+#define platform_longjmp(buf, arg) __builtin_longjmp(buf, arg)
+typedef void *platform_jmp_buf[5];
```
Is this the right size on all architectures?

sunfishcode replied:
[Yes](https://gcc.gnu.org/onlinedocs/gcc/Nonlocal-Gotos.html)

sunfishcode replied:
I've now added a comment about this.

bjorn3 commented:
> An important caveat is that GCC arranges to save and restore only those registers known to the specific architecture variant being compiled for. This can make __builtin_setjmp and __builtin_longjmp more efficient than their library counterparts in some cases, but it can also cause incorrect and mysterious behavior when mixing with code that uses the full register set.

This has me worried. What are the exact conditions in which this causes incorrect behavior?

sunfishcode commented:
> This has me worried. What are the exact conditions in which this causes incorrect behavior?

Suppose one of the hardware architectures we support added a new register, and decides to have a convention that it's compiler-managed, and callee-saved. The host compiler doesn't use the new register, because it's targeting the generic architecture family, but Cranelift does use the new register, because cranelift-native.

Then we could get into a situation like this:

```c
// compiled by cranelift
void outer(void) {
    write_the_new_register(42);
    catcher();
    assert(read_the_new_register() == 42);
}
// compiled by the host compiler
void catcher(void) {
    if (__builtin_setjmp(buf) == 0) {
        inner();
    }
}
// compiled by cranelift
void inner(void) {
    auto save = read_the_new_register();
    clobber_the_new_register();
    if (some condition) {
        jumper();
    }
    write_the_new_register(save);
}
// compiled by the host compiler
void jumper(void) {
    __builtin_longjmp(buf, 1);
}
```
In practice, this should be rare, because the major architectures don't add new registers very often, and when they do, it's usually not compiler-managed, and when it is, it's usually call-clobbered. New compiler-managed callee-saved state also breaks many existing `libc`-implemented `setjmp` functions in the same way, so it's not often done.

froydnj commented:
> > An important caveat is that GCC arranges to save and restore only those registers known to the specific architecture variant being compiled for. This can make __builtin_setjmp and __builtin_longjmp more efficient than their library counterparts in some cases, but it can also cause incorrect and mysterious behavior when mixing with code that uses the full register set.
> 
> This has me worried. What are the exact conditions in which this causes incorrect behavior?

It doesn't require mixing and matching with the full register set.  You run into problems like https://gcc.gnu.org/bugzilla/show_bug.cgi?id=98997 for instance (clang has the same problem).

This particular case is not a problem for wasmtime, because all access to `__builtin_{set,long}jmp` should be within wasmtime itself (and you shouldn't be able to mix-and-match wasmtime runtimes...)

bjorn3 commented:
> In practice, this should be rare, because the major architectures don't add new registers very often

ymm/zmm for avx/avx512 are by default unknown to the compiled wasmtime as only SSE2 is enabled by default for most x86_64 targets.

> and when it is, it's usually call-clobbered.

I see. The x86_64 SystemV does indeed list xmm, ymm and zmm as being clobbered.

sunfishcode commented:
For completeness, a form of this hazard is already present in Wasmtime; if someone statically links Wasmtime, the statically-linked-in `sigsetjmp` will hard-code the registers to save, and if Cranelift at runtime detects and uses more registers and relies on them being callee-saved, it'll also trigger UB.

alexcrichton commented:
We may need to disable this for AArch64 with `clang`, locally when I try to compile it spits out:

```
foo.c:34:7: error: __builtin_setjmp is not supported for the current target
  if (platform_setjmp(buf) != 0) {
      ^~~~~~~~~~~~~~~~~~~~
foo.c:23:30: note: expanded from macro 'platform_setjmp'
#define platform_setjmp(buf) __builtin_setjmp(buf)
                             ^~~~~~~~~~~~~~~~~~~~~
foo.c:44:3: error: __builtin_longjmp is not supported for the current target
  platform_longjmp(*buf, 1);
  ^~~~~~~~~~~~~~~~~~~~~~~~~
foo.c:24:36: note: expanded from macro 'platform_longjmp'
#define platform_longjmp(buf, arg) __builtin_longjmp(buf, arg)
                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~
2 errors generated.
```

Poking around in clang it looks like this is gated on `hasSjLjLowering` which is true for x86 and arm but not for AArch64

sunfishcode commented:
I fixed the code to compile on clang on aarch64, where `__builtin_setmp` isn't supported. If there are other architectures/compilers that don't support it, we can add them as needed.

Talking about this with @cfallin offline, adding new callee-saved registers to cranelift amounts to adding a new ABI, and there are lots of considerations, including calls to and from host code, the register save/restore code in wasmtime-fiber, the `__builtin_setjmp` issue here, trampolines, and possibly other things, so we'll need to carefully think through the design at that time. I've added a comment to the code added in this PR, which seems to be the main useful thing we can do for now.
